refactor: Generalize parameter estimator

jakmeier opened the pull request:
- Change function interfaces to use Gas as a common unit
- Integrate CPU benchmark used to estimate x86 op cost
- Integrate gas metering estimation
- Preparation for more general estimation cases to be added
- Remove unused code

jakmeier commented:
This PR is mostly to give us a base to discuss the direction that I am heading with my refactoring.
This is what I was able to extract as a stable change from a larger set of changes that I have been playing around with.

I guess the main point to discuss here is: How elaborate should the parameter estimator binary be? Should we extend it so that it can run `ICount` and `Time` measurements all in one go? (This is probably necessary to have a meaningful estimation for `Cost::OneCPUInstruction`) Or should we keep it simple and do these comparisons only on the generated cost tables?

If we want to compare different measurements, it might make sense to have multiple more specific calls into Docker+Qemu, instead of one that forwards everything as we have it now. This is part of the motivation for `QemuCommandBuilder`. (The other is that I wanted more debug options.) I could go further and also generalize the docker command.

jakmeier commented on runtime/runtime-params-estimator/src/gas_metering.rs:
```diff
@@ -237,9 +238,16 @@ pub fn compute_gas_metering_cost(
 
-    // println!("with gas: {}; no gas {}", total_raw_with_gas, total_raw_no_gas);
+    // TODO: This seems to fail almost always but has some non-determinism to it.
+    assert!(
```
This is what I am looking at right now with higher priority than the general refactoring, as it seems related to RocksDB compaction. Here. it looks like the IO costs are sometimes attributed to one and sometimes to the other execution. It's always 64kiB for one and 0B for the other. The order in which the calls with/without gas metering happen does not affect it that much. If at least one of the test iterations has the IO cost attributed to the case without gas-metering, then this check fails.

This test also fails in the CI.

matklad replied:
Uhu, let's than maybe comment this out and land just the refactor?

jakmeier replied:
Only commenting the assertion out won't help, the subtraction below will panic with an underflow. But I can add `#[ignore]` to the affected tests

jakmeier replied:
Ah never mind, those tests can be removed anyway

matklad commented on runtime/runtime-params-estimator/src/cost.rs:
```diff
@@ -26,10 +26,6 @@ pub enum Cost {
     ActionDeleteKey,
     ActionDeleteAccount,
 
-    // TODO: Remove these two, compilation is a part of `ActionDeploy`.
-    ContractCompileBase,
-    ContractCompileBytes,
```
I'd actually maybe kept these, as a "helper cost" to answer the question of "how much of deploy cost is compilation, and how much is everything else" . So, move it to the block below, and actually add the relevan functiosn to `lib.rs`

matklad commented on runtime/runtime-params-estimator/src/estimator_params.rs:
```diff
@@ -0,0 +1,55 @@
+// All constant below are measured in Gas, respectively, in fractions thereof.
+
+/// How much gas there is in a nanosecond worth of computation.
+pub(crate) const ONE_NANOSECOND: Ratio<u64> = Ratio::new_raw(1_000_000, 1);
```
```suggestion
pub(crate) const ONE_NANOSECOND: Ratio<Gas> = Ratio::new_raw(1_000_000, 1);
```

matklad commented on runtime/runtime-params-estimator/src/estimator_params.rs:
```diff
@@ -0,0 +1,55 @@
+pub(crate) const ONE_NANOSECOND: Ratio<u64> = Ratio::new_raw(1_000_000, 1);
+// We use factor of 8 to approximately match the price of SHA256 operation between
+// time-based and icount-based metric as measured on 3.2Ghz Core i5.
+pub(crate) const ONE_CPU_INSTRUCTION: Ratio<u64> = Ratio::new_raw(1_000_000, 8);
```
```suggestion
pub(crate) const ONE_CPU_INSTRUCTION: Ratio<Gas> = Ratio::new_raw(1_000_000, 8);
```

matklad commented on runtime/runtime-params-estimator/src/gas_cost.rs:
```diff
@@ -65,8 +67,8 @@ impl GasCost {
-            GasMetric::ICount => result.instructions = raw,
-            GasMetric::Time => result.time_ns = raw,
+            GasMetric::ICount => result.instructions = raw / ONE_CPU_INSTRUCTION,
+            GasMetric::Time => result.time_ns = raw / ONE_NANOSECOND,
```
Maybe name these `GAS_IN_NS` and `GAS_IN_INSTR`?

matklad commented on runtime/runtime-params-estimator/src/gas_metering.rs:
```diff
@@ -25,15 +23,16 @@ fn test_gas_metering_cost(metric: GasMetric) {
-            let cost =
-                compute_gas_metering_cost(metric, VMKind::Wasmer0, REPEATS, &nested_contract);
+            let cost = compute_gas_metering_cost(metric, vm_kind, REPEATS, &nested_contract);
+            #[cfg(test)]
             println!("nested {} {}", depth, cost / (REPEATS as u64));
```
I'd just remove prints

matklad commented on runtime/runtime-params-estimator/src/gas_cost.rs:
```diff
@@ -65,8 +67,8 @@ impl GasCost {
     pub(crate) fn from_raw(raw: Ratio<u64>, metric: GasMetric) -> GasCost {
```
```suggestion
    pub(crate) fn from_raw(raw: Ratio<Gas>, metric: GasMetric) -> GasCost {
```

matklad commented on runtime/runtime-params-estimator/src/config.rs:
```diff
@@ -37,6 +26,6 @@ pub struct Config {
-    /// When non-none, only the specified metrics will be measured.
-    pub metrics_to_measure: Option<Vec<String>>,
+    /// When non-none, only the specified costs will be measured.
+    pub costs_to_measure: Option<Vec<String>>,
```
Not sure if i mentioned this alraedy, but it makes sense to rename the cli flag to jsut `--costs` as well.

matklad commented on runtime/runtime-params-estimator/src/gas_metering.rs:
```diff
@@ -42,29 +41,27 @@ fn test_gas_metering_cost(metric: GasMetric) {
 }
 
 #[test]
 fn test_gas_metering_cost_time() {
```
we moved this to `lib.rs`, so the tests can be safely removed.

matklad reviewed (approved):
LGTM as is, but let's maybe fix the following in this PR:

* drop prints altogether, rather than cfg(test) them
* drop `#[test]` for stuff we moved over to `lib.rs`
* rename constants like `ONE_NANOSECOND` to be more clear

Adding actual separate costs for compilation time and gas can be folow-ups!

matklad commented on runtime/runtime-params-estimator/src/qemu.rs:
```diff
@@ -39,3 +41,72 @@ fn hypercall(index: u32) -> u64 {
+        self
+    }
+    /// Instantiate different counters for each thread
+    pub fn count_per_thread(mut self, count_per_thread: bool) -> Self {
```
I like the following naming convention for such setters (seen in some of BurntSushi's code)

```suggestion
    pub fn count_per_thread(mut self, yes: bool) -> Self {
```

jakmeier commented:
> LGTM as is, but let's maybe fix the following in this PR:
> 
>     * drop prints altogether, rather than cfg(test) them
> 
>     * drop `#[test]` for stuff we moved over to `lib.rs`
> 
>     * rename constants like `ONE_NANOSECOND` to be more clear
> 
> 
> Adding actual separate costs for compilation time and gas can be folow-ups!

Cool, thanks for the quick review! I pushed those changes and also renamed `from_raw` to `from_gas` to make the (new) semantics of that function clear.
